addons/mail/static/src/model/model_core.js

```diff
@@ -2,134 +2,126 @@
+export function patchFields(modelName, fields) {
+    const definition = registry[modelName];
+    if (!definition) {
+        throw new Error(`Model "${modelName}" must be registered before patched.`);
+    }
+    for (const [fieldName, field] of Object.entries(fields)) {
+        if (definition.fields.hasOwnProperty(fieldName)) {
+            throw new Error(`Error patching "${modelName}/${fieldName}": patch to an existing field is not allowed.`);
+        }
+        definition.fields[fieldName] = field;
+    }
 }
 
 /**
- * Register a patch for fields in model.
+ * Applies the provided function to the identifying fields of the model
+ * specified by the `modelName`.
  *
- * @param {string} modelName
- * @param {string} patchName
- * @param {Object} patch
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {function} patch The function to be applied on the identifying fields.
  */
-function registerFieldPatchModel(modelName, patchName, patch) {
-    _registerPatchModel(modelName, patchName, patch, { type: 'field' });
+export function patchIdentifyingFields(modelName, patch) {
+    const definition = registry[modelName];
+    if (!definition) {
+        throw new Error(`Model "${modelName}" must be registered before patched.`);
+    }
+    patch(definition.identifyingFields);
 }
 
 /**
- * Register a patch for identifyignFields in model.
+ * Adds or overrides the provided methods to the model specified by the
+ * `modelName`.
  *
- * @param {string} modelName
- * @param {string} patchName
- * @param {function} patch
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {Object} methods Methods to be added or overriden.
  */
-function registerIdentifyingFieldsPatch(modelName, patchName, patch) {
-    _registerPatchModel(modelName, patchName, patch, { type: 'identifyingFields' });
+export function patchModelMethods(modelName, methods) {
+    const definition = registry[modelName];
+    if (!definition) {
+        throw new Error(`Model "${modelName}" must be registered before patched.`);
+    }
+    if (!definition.modelMethods) {
+        definition.modelMethods = {};
+    }
+    for (const [methodName, method] of Object.entries(methods)) {
+        if (typeof method !== 'function') {
+            throw new Error(`Cannot patch model method on model "${modelName}": ${methodName} is not a function.`);
+        }
+        if (!definition.modelMethods[methodName]) {
+            definition.modelMethods[methodName] = method;
+            continue;
+        }
+        const methodBeforePatch = definition.modelMethods[methodName];
+        definition.modelMethods[methodName] = function(...args) {
+            this._super = methodBeforePatch;
+            return method.call(this, ...args);
+        };
+    }
 }
 
 /**
- * Register a patch for instance methods in model.
+ * Adds or overrides the provided methods to the records of the model specified
```

Comment:
I think we should have different methods (and different checks in place) whether the user is trying to add or override.

Reply:
The idea is good, works perfectly with the current codebase, and even helped me to detect some tricky errors, but what about this hypothetical case?:

- `mail.demo` has no fields
- `voip` module needs the field `test` to be added to `mail.demo`
- `im_livechat` module as well

In this case, whether the field should be patched or added will depend on which file is loaded first, and this sounds pretty bad.